diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -503,75 +506,190 @@ int DifferentialFunctionComparator::cmpAllocs(const CallInst *CL,
            || STyL->getStructName() != STyR->getStructName();
 }
 
-/// Check if the given operation can be ignored (it does not affect semantics)
-/// for control flow only diffs.
-bool DifferentialFunctionComparator::mayIgnore(const User *Inst) const {
-    if (config.ControlFlowOnly)
-        return isa<AllocaInst>(Inst) || isCast(Inst);
-    else {
-        if (isa<AllocaInst>(Inst))
+/// Check if the given instruction can be ignored (it does not affect
+/// semantics) for control flow only diffs. Replacements of ignorable
+/// instructions are stored inside the ignored instructions map.
+bool DifferentialFunctionComparator::maySkipInstruction(
+        const Instruction *Inst) const {
+    if (isa<AllocaInst>(Inst)) {
+        // Ignore AllocaInsts with no specific replacement.
+        return true;
+    }
+    if (config.ControlFlowOnly) {
+        if (isCast(Inst)) {
+            ignoredInstructions.insert({Inst, Inst->getOperand(0)});
             return true;
+        }
+    } else {
         if (isCast(Inst)) {
-            auto SrcTy = Inst->getOperand(0)->getType();
-            auto DestTy = Inst->getType();
-
-            if (auto StrTy = dyn_cast<StructType>(SrcTy)) {
-                if (StrTy->hasName() && StrTy->getName().startswith("union"))
-                    // Casting from a union type is safe to ignore, because in
-                    // case it was generated by something different than a
-                    // replacement of a simple type with a union, the
-                    // comparator would detect a difference when the value is
-                    // used, since it would have a different type.
-                    return true;
-            }
+            return maySkipCast(Inst);
+        }
+        if (auto Load = dyn_cast<LoadInst>(Inst)) {
+            return maySkipLoad(Load);
+        }
+    }
```

Comment:
I think that a better style of this would be:

    if (isCast(Inst)) {
        if (config.ControlFlowOnly)
            ...
        else
            ...
    }
    if (auto Load = ...) {
        ...
    }

to have all the top-level if conditions checking the instruction kind.